src/linear_algebra/Hermitian_inner_product.lean

```diff
@@ -0,0 +1,669 @@
+  cases ho,
+  { rw ho,
+    simp },
+
+  { rw [←of_real_mul, of_real_herm_norm_mul_self,
+      div_self (inprod_self_ne_zero_iff.mpr ho), one_smul] }
+end
+
+lemma smul_proj (x y : α) {a : ℂ} : 
+proj (a • x) y = a • (proj x y) :=
+begin
+  dunfold proj,
+  rw [inprod_smul_left, smul_smul],
+  ring,
+end
+
+lemma proj_smul (x y : α) {a : ℂ} (ha : a ≠ 0) :
+proj x (a • y) = proj x y := 
+begin
+  have Hy : y = 0 ∨ y ≠ 0,
+  { apply em },
+  cases Hy,
+  { rw [Hy, smul_zero] },
+
+  { dunfold proj,
+    rw [inprod_smul_right, herm_norm_smul, of_real_mul, smul_smul],
+    suffices :  ((↑(abs a) * ↑|y| * (↑(abs a) * ↑|y| ))⁻¹ * x ₀ y * conj a * a) • y = ((↑|y| * ↑|y| )⁻¹ * x ₀ y) • y,
+    { rw [div_eq_inv_mul, div_eq_inv_mul,
+        field.mul_comm (conj a), ←field.mul_assoc],
+      exact this },
+    { rw [field.mul_assoc, ←field.mul_comm a, mul_conj, 
+        ←field.mul_assoc ( ↑(abs a) * ↑|y| ), field.mul_comm (↑(abs a)),
+        field.mul_assoc ( ↑|y| ), ←of_real_mul, mul_self_abs,
+        field.mul_comm ( ↑|y| ), field.mul_comm, field.mul_assoc,
+        mul_inv_eq, field.mul_comm, field.mul_assoc ((↑|y| * ↑|y| )⁻¹),
+        field.mul_comm (↑(norm_sq a))⁻¹, field.mul_assoc, field.mul_assoc (x ₀ y),
+        field.inv_mul_cancel, field.mul_one],
+      refl,
+      
+      exact of_real_ne_zero.mpr ((iff_false_right ha).mp (norm_sq_eq_zero)),
+      exact (mul_ne_zero (of_real_ne_zero.mpr ((norm_ne_zero_iff_ne_zero).mpr Hy)) (of_real_ne_zero.mpr ((norm_ne_zero_iff_ne_zero).mpr Hy))),
+      exact of_real_ne_zero.mpr ((iff_false_right ha).mp (norm_sq_eq_zero)) }}
+end
+
+lemma norm_proj_le (x y : α) : 
+|proj x y| ≤ |x| :=
+begin
+  have Hy : y = 0 ∨ y ≠ 0,
+  { apply em },
+  cases Hy,
+  { rw [Hy, proj_zero, herm_norm_zero],
+    exact @norm_nonneg _ _ x },
+
+  { dunfold proj,
+    rw [herm_norm_smul, complex.abs_div, ←of_real_mul,
+      abs_of_real, abs_mul_self, div_mul_comm, division_def,
+      mul_inv', ←field.mul_assoc, field.mul_inv_cancel,
+      field.one_mul, field.mul_comm, ←division_def, div_le_iff],
+     
+      exact abs_inprod_le_mul_herm_norm x y,
+      exact (norm_pos_iff y).mpr Hy,
+      exact norm_ne_zero_iff_ne_zero.mpr Hy }
+end
+
+lemma ortho_imp_proj_eq_zero {x y : α} :
+(x ⊥ y) → proj x y = 0 := 
+begin
+  dunfold herm_ortho,
+  dunfold proj,
+  intros H,
+  rw H,
+  simp,
+end
+
+lemma proj_eq_self_iff_lindep {x y : α} :
+proj x y = x ↔ ∃ (a : ℂ), x = a • y :=
+begin
+  split,
+  { dunfold proj, 
+    intros H,
+    exact exists.intro (x ₀ y / (↑|y| * ↑|y| )) (eq_comm.mp H) },
+
+  { intros H,
+    cases H with a Ha,
+    rw [Ha, smul_proj, proj_self_eq_self] }
+end
+
+end herm_inner_product_space
+
+class Hilbert_space (α : Type*) [add_comm_group α] [vector_space ℂ α] extends herm_inner_product_space α :=
+(complete : ∀{f : filter α}, cauchy f → ∃x, f ≤ nhds x) 
+
+instance Hilbert_space.to_complete_space (α : Type*) [add_comm_group α] [vector_space ℂ α] [Hilbert_space α] : complete_space α :=
```

Comment:
Can't you just make `Hilbert_space` extend `herm_inner_product_space` and `complete_space` by definition?